samples/bluetooth/nrf_dm/src/peer.c

```diff
@@ -0,0 +1,398 @@
+		res = peer->result.dist_estimates.rtt.rtt;
+	} else {
+		res = peer->result.dist_estimates.mcpd.best;
+	}
+
+	res = (res < 0 ? 0 : res) * 10;
+	if (res > DISTANCE_MAX_LED) {
+		pwm_led_set(0);
+	} else {
+		pwm_led_set((UINT16_MAX/DISTANCE_MAX_LED * res * (-1)) + UINT16_MAX);
+	}
+}
+
+static void print_result(struct dm_result *result)
+{
+	if (!result) {
+		return;
+	}
+
+	const char *quality[DM_QUALITY_NONE + 1] = {"ok", "poor", "do not use", "crc fail", "none"};
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(&result->bt_addr, addr, sizeof(addr));
+
+	printk("\nMeasurement result:\n");
+	printk("\tAddr: %s\n", addr);
+	printk("\tQuality: %s\n", quality[result->quality]);
+
+	printk("\tDistance estimates: ");
+	if (result->ranging_mode == DM_RANGING_MODE_RTT) {
+		printk("rtt: rtt=%.2f\n", result->dist_estimates.rtt.rtt);
+	} else {
+		printk("mcpd: ifft=%.2f phase_slope=%.2f rssi_openspace=%.2f best=%.2f\n",
+			result->dist_estimates.mcpd.ifft,
+			result->dist_estimates.mcpd.phase_slope,
+			result->dist_estimates.mcpd.rssi_openspace,
+			result->dist_estimates.mcpd.best);
+	}
+}
+
+static void timeout_handler(struct k_timer *timer_id)
+{
+	sys_snode_t *node, *tmp;
+	struct peer_entry *item;
+
+	SYS_SLIST_FOR_EACH_NODE_SAFE(&peer_list, node, tmp) {
+		item = CONTAINER_OF(node, struct peer_entry, node);
+		if (item->timeout_ms > PEER_TIMEOUT_STEP_MS) {
+			item->timeout_ms -= PEER_TIMEOUT_STEP_MS;
+		} else {
+			sys_slist_remove(&peer_list, NULL, node);
+			k_heap_free(&peer_heap, item);
+			closest_peer = peer_find_closest();
+
+			led_notification(closest_peer);
+		}
+	}
+}
+
+static void peer_thread(void)
+{
+	struct dm_result result;
+
+	while (1) {
+		if (k_msgq_get(&result_msgq, &result, K_FOREVER) == 0) {
+			struct peer_entry *peer;
+
+			peer = peer_find(&result.bt_addr);
+			if (!peer) {
+				continue;
+			}
+
+			memcpy(&peer->result, &result, sizeof(peer->result));
+			peer->timeout_ms = PEER_TIMEOUT_INIT_MS;
+			print_result(&peer->result);
+
+			closest_peer = mcpd_min_peer_result(closest_peer, peer);
+
+			led_notification(closest_peer);
+			ble_notification(closest_peer);
+		}
+	}
+}
+
+bool peer_supported_test(const bt_addr_le_t *peer)
+{
+	sys_snode_t *node, *tmp;
+	struct peer_entry *item;
+
+	SYS_SLIST_FOR_EACH_NODE_SAFE(&peer_list, node, tmp) {
+		item = CONTAINER_OF(node, struct peer_entry, node);
+		if (bt_addr_le_cmp(&item->bt_addr, peer) == 0) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+int peer_supported_add(const bt_addr_le_t *peer)
+{
+	struct peer_entry *item;
+
+	if (peer_supported_test(peer)) {
+		return 0;
+	}
+
+	item = k_heap_alloc(&peer_heap, sizeof(struct peer_entry), K_NO_WAIT);
+	if (!item) {
+		return -ENOMEM;
+	}
+
+	item->timeout_ms = PEER_TIMEOUT_INIT_MS;
+	bt_addr_le_copy(&item->bt_addr, peer);
+	list_lock();
+	sys_slist_append(&peer_list, &item->node);
+	list_unlock();
+
+	return 0;
+}
+/** Prepare access address as per BT Spec.
```

Comment:
This code comes from zephyr which is not licensed under the nordic license. Maybe we need to check if copying it here is ok.

Reply:
I have changed the method of preparing the access address for dm.